api/swagger.yaml

```diff
@@ -9900,7 +9900,9 @@ paths:
               Id: "22be93d5babb089c5aab8dbc369042fad48ff791584ca2da2100db837a1c7c30"
               Warning: ""
         403:
-          description: "operation not supported for pre-defined networks"
+          description: |
+            Forbidden operation. This happens when trying to create a network named after a pre-defined network,
+            or when trying to create an overlay network on a daemon which is not part of a Swarm cluster.
```

Comment:
I think a 400 error would actually be more appropriate here. Not sure why a 403 was picked before, but it doesn't look like the correct error;  https://www.codetinkerer.com/2015/12/04/choosing-an-http-status-code.html

Reply:
I think both status code could fit here. From [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html):

> The 400 (Bad Request) status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).

> The 403 (Forbidden) status code indicates that the server understood the request but refuses to fulfill it. A server that wishes to make public why the request has been forbidden can describe that reason in the response content (if any). [...] **However, a request might be forbidden for reasons unrelated to the credentials.**

For both cases, the request is well-formed and the issue doesn't come from the client request per se, but the daemon refuses the request because this action is forbidden.

On the other hand, for the pre-defined networks case, one could compare this a duplicate error. But `CheckDuplicate` also triggers an (undocumented) 403.

> Or are we using 403 specifically for swarm operations against a non-swarm node?

* `/networks/{id}/connect` and `/networks/{id}/disconnect` returns a 403 for swarm-scoped networks ;
* `/services/create`, `/volumes/{name}`, `/nodes/...`, `/swarm/...` and `/services/...` routes return a 503 when node isn't part of a swarm cluster ;

Reply:
Yeah, 403 is not "invalid" to return, but not very common in our API (and probably more commonly used for Authentication issues), which is why I was wondering if it would make sense to change it to a 400.

Reply:
I find 503 to be an odd choice. I agree that 403 is a better fit for the "can't do unless daemon is part of a swarm" error. I also think [501 Not Implemented](https://www.rfc-editor.org/rfc/rfc9110.html#name-501-not-implemented) (with a `Cache-control: no-store` response header) could also be a good fit as the error is more a problem with the state of the server than with the identity of the client or content of the request. That being said, I think 403 is the better choice only because of the precedent set by `/networks/{id}/connect`.

Reply:
Yeah, the 503 isn't great. I recall there was some debate about the best option for that; a 404 would technically have been correct (as the endpoints don't exist if swarm is not enabled), but that would also be somewhat confusing, so ultimately it converged to 503.